### ISSUE TITLE
There are some issues in network layer

### ISSUE THREAD

**Issue by luongvo** [anticafe/survey_app#21]
- OkHttp Authenticator should have a limit time of retry or a condition to break, otherwise, there would be a loop in renewing token while the refreshToken is invalidated (maybe password changed, token invoked from the backend). See https://square.github.io/okhttp/recipes/#handling-authentication-kt-java
- Storing the token to local storage shouldn't be done with a whole entity, I wonder what would happen if we change the entity structure? Proper structure for storage is better to have. Why don’t we use Room also or just store access token value instead of the whole entity?
- `SurveyLocalDataSource` is not being used anywhere?
- Quite many generic Exception catches (e.g: https://github.com/anticafe/survey_app/blob/development/app/src/main/java/com/myapp/data/repo/AccountRepository.kt#L35)
- Something isn't correct here: https://github.com/anticafe/survey_app/blob/4c79d8fc64347199a6648c19d576425d2ab8b694/app/src/main/java/com/myapp/data/repo/SurveyRepository.kt#L39 the parsing looks potentially duplicated in case we have another API call.
- Hardcoded message, instead, just return sth like an "EmptyBodyDataException", because then the downstream can check for the type and handle it accordingly, also, error message should not be processed here at this layer, it should be pure logic/error code processor only, showing message to users should be handled on the higher layer i.e View, VM https://github.com/anticafe/survey_app/blob/4c79d8fc64347199a6648c19d576425d2ab8b694/app/src/main/java/com/myapp/data/repo/RepoExceptionUtil.kt#L12

Things could be improved as well:
- Dispatching the network request could be better handled with RxJava or Coroutine to enhance the concurrency calls when later, the project could be grown-up extensively.

**Comment by trietbui85** [comment]
This issue has been fixed in dd25217
>OkHttp Authenticator should have a limit time of retry

**Comment by trietbui85** [comment]
@luongvo : About handle exception:
>Hardcoded message, instead, just return sth like an "EmptyBodyDataException", because then the downstream can check for the type and handle it accordingly, also, error message should not be processed here at this layer, it should be pure logic/error code processor only, showing message to users should be handled on the higher layer i.e View, VM
>https://github.com/anticafe/survey_app/blob/4c79d8fc64347199a6648c19d576425d2ab8b694/app/src/main/java/com/myapp/data/repo/RepoExceptionUtil.kt#L12

Here is my thought about exception handle (and what I've already implemented before submitting code):
- `DataException(val httpCode: Int, val errorMessage: String)` will be use to unify all exceptions from repository layer. Although it have `httpCode` and `errorMessage` but the second one is for debug purpose. That's why in [RepoExceptionUtil](https://github.com/anticafe/survey_app/blob/development/app/src/main/java/com/myapp/data/repo/RepoExceptionUtil.kt), I'm really hardcode exception text if no text available.
- In case hardcode error text like "No response body", in fact I don't care about that message because the `httpCode` from Response will be the key factor to determine what's the root cause of error. Thus I don't think something like `EmptyBodyDataException` is needed in this case.
- In presentation layer (Fragment or ViewModel), we will convert that `DataException.httpCode` into real error message (via method `getErrorText` in [ExceptionExt](https://github.com/anticafe/survey_app/blob/development/app/src/main/java/com/myapp/utils/ExceptionExt.kt)). In Fragment side, it looks like this:
https://github.com/anticafe/survey_app/blob/2707bcfa29258f8f1a5e38d0611a464260a7c0d9/app/src/main/java/com/myapp/ui/main/MainFragment.kt#L116-L122

It seems at the first time review the RepoExceptionUtil, you thought it doesn't work like I explain. Please correct me if I'm wrong. Your comments are welcome 👍

**Comment by trietbui85** [comment]
@luongvo : About generic exception catch:
>Quite many generic Exception catches (e.g: https://github.com/anticafe/survey_app/blob/development/app/src/main/java/com/myapp/data/repo/AccountRepository.kt#L35)

There are 3 places which has generic exception:
- RepoDataMapper: Thanks for let me know 👏  . Fixed in 2c1f4e4 by changing exception type to JsonSyntaxException
- SurveyRepository: https://github.com/anticafe/survey_app/blob/2707bcfa29258f8f1a5e38d0611a464260a7c0d9/app/src/main/java/com/myapp/data/repo/SurveyRepository.kt#L39-L44
- AccountRepository: https://github.com/anticafe/survey_app/blob/2707bcfa29258f8f1a5e38d0611a464260a7c0d9/app/src/main/java/com/myapp/data/repo/AccountRepository.kt#L35

But, I still think we should keep the generic exception in repository. Notice that we don't handle exception in local (Room) or remote (Retrofit) module, thus all exceptions will be handled in repository module instead. Imagine there are 5 exceptions from remote, and 5 from local, there will be 10 to catch here. It would make code longer with 10 `try {} catch {}`. And, how about when you change the implementation of local, which throw a new kind of exception - in this case if you must update this place again to catch that exception.
What I do here is use generic exception and let `e.parseConnectionException()` to convert that exception to our DataException. All exception in one place.

How do you think about my approach? Your comments are welcome 👍

**Comment by trietbui85** [comment]
About token storage:
>Storing the token to local storage shouldn't be done with a whole entity, I wonder what would happen if we change the entity structure? Proper structure for storage is better to have. Why don’t we use Room also or just store access token value instead of the whole entity?

If you're worried when the structure of token is changed in server side, then I think the entity is still be fine. Because current flow of token data is TokenResonse (remote) -> TokenItem (repo) -> TokenEntity (local). In each step, the [AccountItemMapper](https://github.com/anticafe/survey_app/blob/development/app/src/main/java/com/myapp/data/repo/RepoDataModel.kt) will be in charge on convert item to another item. With this approach (borrow from Clean Architecture), each module (remote, local, repo) is independent, and changes (for ex: model structure) from module remote will not affect to module local.

I also save the whole entity (with accessToken, tokenType, expiresIn, createdAt) to cache instead just accessToken because I'm not sure we will use expiresIn or createdAt later (for example we might need to check whether token expires in client side?). So, better to save them all.

Use SharedPreferences instead Room because SharedPreferences looks more flexible, and better for something small like Token. We even can use [EncryptedSharedPreferences](https://developer.android.com/reference/androidx/security/crypto/EncryptedSharedPreferences) to encrypt sensitive data like token.

@luongvo Do you think that is fine?

**Comment by trietbui85** [comment]
>SurveyLocalDataSource is not being used anywhere?

At first I follow repository pattern, thus set up both local and remote DataSource. But later, when I look at the Technical Test document again, there is no requirement to load survey from local DataSource first. Since I don't have your contact for asking to make it clear, I still keep the SurveyLocalDataSource there. 

So, @luongvo , my question is: do we need to load survey data from cache first? For better user experience, seeing some cache surveys displayed instantly is preferred than looking at the progress loading and waiting for 2 seconds.

**Comment by trietbui85** [comment]
About this improvement:
>Dispatching the network request could be better handled with RxJava or Coroutine to enhance the concurrency calls when later, the project could be grown-up extensively.

I make a PR #32 for it. @luongvo Could you help me to review it? (Not sure why I cannot add you as reviewer for that PR). Thanks.

**Comment by luongvo** [comment]
> So, @luongvo , my question is: do we need to load survey data from cache first? For better user experience, seeing some cache surveys displayed instantly is preferred than looking at the progress loading and waiting for 2 seconds.

The thing is there are some things redundant in your code and that makes us can't catch up with your idea clearly. Please try to manage the code, let's remove unnecessary, redundant things out of the code since we had git for version control. 🤷‍♂

**Comment by luongvo** [comment]
> Use SharedPreferences instead Room because SharedPreferences looks more flexible, and better for something small like Token. We even can use EncryptedSharedPreferences to encrypt sensitive data like token.

Agree, but the combination of SharedPreferences and Gson is not a best practice way to follow imo :)

**Comment by trietbui85** [comment]
>Please try to manage the code, let's remove unnecessary, redundant things out of the code since we had git for version control. 🤷‍♂

Fixed in 7326f50

**Comment by trietbui85** [comment]
>Agree, but the combination of SharedPreferences and Gson is not a best practice way to follow imo :)

I make commit 8e640c6 to refactor the TokenLocalDataSource by only save accessToken to cache instead whole object. Do you think it's okay now @luongvo ?
